### PR TITLE
Collect hash table sizes statistics in ConcurrentHashJoin

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Sizes of hash tables created by join (`parallel_hash` algorithm) is collected and cached now. This information will be used to preallocate space in hash tables for subsequent query executions and save time on hash table resizes.

---

Screenshot 2024-06-11 at 13 07 18

### CODE REVIEW

**Comment by nickitat** [comment]
based on https://github.com/ClickHouse/ClickHouse/pull/64241 and https://github.com/ClickHouse/ClickHouse/pull/64550, so need to wait until those are merged

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit cc0c53824e6d39607d5116cf8205a3e552ace97f with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64553/cc0c53824e6d39607d5116cf8205a3e552ace97f/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Planner/PlannerJoins.cpp
Comment (devcrafter) [inline src/Planner/PlannerJoins.cpp:813]: Let's use separate settings for join. To see `max_size_to_preallocate_for_aggregation` here is confusing

**Review by devcrafter — APPROVED** [review]
LGTM with comments

**Pull request merged**
